js/modules/loki_primitives.js

```diff
@@ -0,0 +1,135 @@
+/* global clearTimeout */
+// was timeoutDelay
+const sleepFor = ms => new Promise(resolve => setTimeout(resolve, ms));
+
+let log;
+function configure(options = {}) {
+  ({ log } = options);
+}
+
+// Taken from https://stackoverflow.com/questions/51160260/clean-way-to-wait-for-first-true-returned-by-promise
+// The promise returned by this function will resolve true when the first promise
+// in ps resolves true *or* it will resolve false when all of ps resolve false
+const firstTrue = ps => {
+  const newPs = ps.map(
+    p =>
+      new Promise(
+        // eslint-disable-next-line more/no-then
+        (resolve, reject) => p.then(v => v && resolve(v), reject)
+      )
+  );
+  // eslint-disable-next-line more/no-then
+  newPs.push(Promise.all(ps).then(() => false));
+  return Promise.race(newPs);
+};
+
+// one action resolves all
+const snodeGlobalLocks = {};
+async function allowOnlyOneAtATime(name, process, timeout) {
+  // if currently not in progress
+  if (snodeGlobalLocks[name] === undefined) {
+    // set lock
+    snodeGlobalLocks[name] = new Promise(async (resolve, reject) => {
+      // set up timeout feature
+      let timeoutTimer = null;
+      if (timeout) {
+        timeoutTimer = setTimeout(() => {
+          log.warn(
+            `loki_snodes:::allowOnlyOneAtATime - TIMEDOUT after ${timeout}s`
+          );
+          delete snodeGlobalLocks[name]; // clear lock
+          reject();
+        }, timeout);
+      }
+      // do actual work
+      let innerRetVal;
+      try {
+        innerRetVal = await process();
+      } catch (e) {
+        log.error(
+          `loki_snodes:::allowOnlyOneAtATime - error ${e.code} ${e.message}`
+        );
+        throw e;
```

Comment:
Don't we need to clean-up here (clearTimeout, clear lock)?

Reply:
yup